cli/flox-rust-sdk/src/models/lockfile.rs

```diff
@@ -159,47 +278,181 @@ impl LockedManifestCatalog {
+            .map_err(LockedManifestError::CatalogResolve)?;
+
+        let locked_packages = Self::locked_packages_from_resolution(manifest, resolved).collect();
+
+        let lockfile = LockedManifestCatalog {
+            version: Version::<1>,
+            manifest: manifest.clone(),
+            packages: locked_packages,
+        };
+
+        Ok(lockfile)
+    }
+
+    /// Transform a lockfile into a mapping  that is easier to query:
+    /// Lockfile -> { (package, system): locked package }
+    fn make_seed_mapping(
+        seed: &LockedManifestCatalog,
+    ) -> HashMap<(&ManifestPackageDescriptor, &System), &LockedPackageCatalog> {
+        seed.packages
+            .iter()
+            .filter_map(|package| {
+                let system = &package.system;
+                let manifest = seed.manifest.install.get(&package.name)?;
+                Some(((manifest, system), package))
+            })
+            .collect()
+    }
+
+    /// Creates package groups from a flat map of install descriptors
+    ///
+    /// A group is created for each unique combination of (descriptor.package_group ｘ descriptor.system).
+    /// Each group contains a list of package descriptors that belong to that group.
+    ///
+    /// `locked` is used to provide existing derivations for packages that are already locked,
+    /// e.g. by a previous lockfile.
+    /// These packages are used to constrain the resolution.
+    fn collect_package_groups<'manifest>(
+        manifest: &'manifest TypedManifestCatalog,
+        seed_lockfile: Option<&LockedManifestCatalog>,
+    ) -> impl Iterator<Item = PackageGroup> + 'manifest {
+        let seed_locked_packages = seed_lockfile.map_or_else(HashMap::new, Self::make_seed_mapping);
+
+        // Using a btree map to ensure consistent ordering
+        let mut map = BTreeMap::new();
+
+        let default_systems = &manifest.options.systems;
+
+        for (install_id, manifest_descriptor) in manifest.install.iter() {
+            let resolved_descriptor = PackageDescriptor {
+                name: install_id.clone(),
+                pkg_path: manifest_descriptor.pkg_path.clone(),
+                derivation: None,
+                semver: None,
+                version: manifest_descriptor.version.clone(),
+            };
+
+            let group = manifest_descriptor
+                .package_group
+                .as_deref()
+                .unwrap_or(DEFAULT_GROUP_NAME);
+
+            let descriptor_systems = manifest_descriptor
+                .systems
+                .as_ref()
+                .unwrap_or(default_systems);
+
+            for system in descriptor_systems {
+                let resolved_group = map.entry((group, system)).or_insert_with(|| PackageGroup {
+                    descriptors: Vec::new(),
+                    name: group.to_string(),
+                    system: system.clone(),
+                });
+
+                let locked_derivation = seed_locked_packages
```

Comment:
nit
```suggestion
                // If the package was just installed, this will be `None`
                let locked_derivation = seed_locked_packages
```